lightning/backend/sql2kv.go

```diff
@@ -65,15 +67,27 @@ func NewTableKVEncoder(tbl table.Table, options *SessionOptions) (Encoder, error
 	recordCtx := tables.NewCommonAddRecordCtx(len(cols))
 	tables.SetAddRecordCtx(se, recordCtx)
 
-	var autoRandomBits int64
+	autoIDFn := func(id int64) int64 { return id }
 	if meta.PKIsHandle && meta.ContainsAutoRandomBits() {
 		for _, col := range cols {
 			if mysql.HasPriKeyFlag(col.Flag) {
 				incrementalBits := autoRandomIncrementBits(col, int(meta.AutoRandomBits))
-				autoRandomBits = rand.New(rand.NewSource(options.AutoRandomSeed)).Int63n(1<<meta.AutoRandomBits) << incrementalBits
+				autoRandomBits := rand.New(rand.NewSource(options.AutoRandomSeed)).Int63n(1<<meta.AutoRandomBits) << incrementalBits
+				autoIDFn = func(id int64) int64 {
+					return autoRandomBits | id
+				}
 				break
 			}
 		}
+	} else if meta.ShardRowIDBits > 0 {
+		rd := rand.New(rand.NewSource(options.AutoRandomSeed))
+		mask := int64(1)<<meta.ShardRowIDBits - 1
+		shift := autoid.RowIDBitLength - meta.ShardRowIDBits - 1
+		autoIDFn = func(id int64) int64 {
+			rd.Seed(id)
+			shardBits := (int64(rd.Uint32()) & mask) << shift
+			return shardBits | id
+		}
```

Comment:
How about using TiDBs `OverflowShardBits`?

Reply:
Seems `OverflowShardBits` used to check the shard row id is overflow. But here we just want to generate a random shardBits that bind with the lightning generated rowid so it's compatible with checkpoint.